tests/shadow/test_shadow_class.py

```diff
@@ -0,0 +1,474 @@
+        wires = range(2)
+
+        dev = qml.device("default.qubit", wires=wires, shots=10000)
+
+        @qml.qnode(dev)
+        def qnode(n_wires):
+            qml.Hadamard(0)
+            qml.CNOT(wires=[0, 1])
+            return qml.classical_shadow(wires=range(n_wires))
+
+        # should prepare the bell state
+        bits, recipes = qnode(2)
+        shadow = ClassicalShadow(bits, recipes)
+        global_snapshots = shadow.global_snapshots()
+
+        state = np.sum(global_snapshots, axis=0) / shadow.snapshots
+        bell_state = np.array([[0.5, 0, 0, 0.5], [0, 0, 0, 0], [0, 0, 0, 0], [0.5, 0, 0, 0.5]])
+        assert qml.math.allclose(state, bell_state, atol=1e-1)
+
+        # reduced state should yield maximally mixed state
+        local_snapshots = shadow.local_snapshots(wires=[0])
+        assert qml.math.allclose(np.mean(local_snapshots, axis=0)[0], 0.5 * np.eye(2), atol=1e-1)
+
+        # alternative computation
+        bits, recipes = qnode(1)
+        shadow = ClassicalShadow(bits, recipes)
+        global_snapshots = shadow.global_snapshots()
+        local_snapshots = shadow.local_snapshots(wires=[0])
+
+        state = np.sum(global_snapshots, axis=0) / shadow.snapshots
+        assert qml.math.allclose(state, 0.5 * np.eye(2), atol=1e-1)
+
+
+def hadamard_circuit(wires, shots=10000, interface="autograd"):
+    """Hadamard circuit to put all qubits in equal superposition (locally)"""
+    dev = qml.device("default.qubit", wires=wires, shots=shots)
+
+    @qml.qnode(dev, interface=interface)
+    def circuit():
+        for i in range(wires):
+            qml.Hadamard(wires=i)
+        return qml.classical_shadow(wires=range(wires))
+
+    return circuit
+
+
+def max_entangled_circuit(wires, shots=10000, interface="autograd"):
+    """maximally entangled state preparation circuit"""
+    dev = qml.device("default.qubit", wires=wires, shots=shots)
+
+    @qml.qnode(dev, interface=interface)
+    def circuit():
+        qml.Hadamard(wires=0)
+        for i in range(1, wires):
+            qml.CNOT(wires=[0, i])
+        return qml.classical_shadow(wires=range(wires))
+
+    return circuit
+
+
+def qft_circuit(wires, shots=10000, interface="autograd"):
+    """Quantum Fourier Transform circuit"""
+    dev = qml.device("default.qubit", wires=wires, shots=shots)
+
+    one_state = np.zeros(wires)
+    one_state[-1] = 1
+
+    @qml.qnode(dev, interface=interface)
+    def circuit():
+        qml.BasisState(one_state, wires=range(wires))
+        qml.QFT(wires=range(wires))
+        return qml.classical_shadow(wires=range(wires))
+
+    return circuit
+
+
+# marked slow because state reconstruction for high number of qubits is slow
+@pytest.mark.slow
+@pytest.mark.all_interfaces
+class TestStateReconstruction:
+    """Test that the state reconstruction is correct for a variety of states"""
+
+    @pytest.mark.parametrize("wires", [1, 3])
+    @pytest.mark.parametrize("interface", ["autograd", "jax", "tf", "torch"])
```

Comment:
Do you think we could only test one reconstruction with all interfaces, and then do all the other reconstructions with a fixed interface? We would save a lot of slow tests.

If the tests cover different parts of the src code then we can leave it as it is.

Reply:
I think that might be a good compromise solution for reducing test runtime 👍

Reply:
Done.